Issue 3: historical price response storage

asmi38 opened the pull request:
### Issue 3: Add Response Storage to Historical Price Endpoint
The Historical Prices endpoint currently forwards calls to the Iex Historical Prices API every time, even if the same data is requested twice. We should speed up subsequent calls to the FUSE API by storing previously requested Historical Prices, serving them to the caller without making an external call to the Iex API.

asmi38 committed:
Setup application settings for h2 in-mem db

Adjusted the h2 db url, enabled the console and changed its path, enabled logging of JPA queries to standard output.

asmi38 committed:
Created HistoricalPrice components for the Entity, Repository and Primary key

Created IexHistoricalPriceEntity which serves as the framework for the storage of historical prices.

Created IexHistoricalPriceId which serves as a class to enable the use of a composite primary key of the symbol and date fields.

Created IexHistoricalPriceRpsy which enables the use of CRUD operations on the table.

asmi38 committed:
Added response storage to IEX queries

Added method to IexService which adds the prices retrieved from IEX to the database.

Added naming to IexHistoricalPriceEntity

alex-power commented on src/main/java/org/galatea/starter/domain/IexHistoricalPriceId.java:
```diff
@@ -0,0 +1,33 @@
+  private String date;
+
+  @Override
+  public boolean equals(final Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    IexHistoricalPriceId iexHistoricalPriceId = (IexHistoricalPriceId) o;
+    return symbol.equals(iexHistoricalPriceId.symbol) && date.equals(iexHistoricalPriceId.date);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(symbol, date);
+  }
```
Equals and HashCode functions can be automatically generated by Lombok using the `@EqualsAndHashCode` annotation, alternatively use `@Data` which includes `@EqualsAndHashCode` as well as `@ToString`, `@Getters`, `@Setters`, and `@RequiredArgsConstructor`.

alex-power commented on src/main/java/org/galatea/starter/service/IexService.java:
```diff
@@ -58,14 +62,38 @@
+          price.getLow(),
+          price.getOpen(),
+          price.getVolume());
+      historicalPriceRspy.save(entity);
```
Using saveAll would be more efficient due to the way hibernate will execute a save, but it's not too important here.

alex-power reviewed (changes requested):
I don't see any code to recall the relevant saved data from the DB and insert it in the response. So this is just the storage part?

asmi38 committed:
Added date checker class

Date checkers class provides functions to check if a specific date is a weekend or a trading holiday for the past 5 years

asmi38 committed:
Added functionality for storage and optimization of queries to the IEX API

Added functionality for storage and optimization of queries to the IEX API. The current implementation of determining the best query is naive and can be improved if a better cost of queries can be established.

grundygolem2 commented on src/main/java/org/galatea/starter/domain/IexHistoricalPrice.java:
```diff
@@ -1,10 +1,13 @@
 import lombok.Data;
 
 @Data
+@AllArgsConstructor(access = AccessLevel.PUBLIC)
```
I think access level defaults to public

grundygolem2 commented on src/main/java/org/galatea/starter/utils/DateChecker.java:
```diff
@@ -2,84 +2,81 @@
-        "2015-12-25"));
-  }
+public final class DateChecker {
+  static Set<String> holidayDates = Stream.of("2021-01-01",
```
You should not hardcode lists of dates like this. What are some better ways to determine this?

asmi38 replied:
I found a way of calculating the weekends, but couldn't figure out an efficient way of determining exchange holidays in a cleaner way - mostly because there was a lot of variation from year to year.
